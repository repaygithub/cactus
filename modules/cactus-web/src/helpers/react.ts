import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
