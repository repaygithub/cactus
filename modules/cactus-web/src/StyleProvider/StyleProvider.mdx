---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
