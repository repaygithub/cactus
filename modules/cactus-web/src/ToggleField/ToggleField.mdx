---
name: ToggleField
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import ToggleField from './ToggleField'
import cactusTheme from '@repay/cactus-theme'
import React, { useState } from 'react'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# ToggleField

### Try it out

export const code = `() => {
  const [state, setState] = useState(true)
  return (
    <ToggleField
      name="enabled"
      label={state ? 'On' : 'Off'}
      checked={state}
      onChange={(e) => setState(e.target.checked)}
    />
  )
}`

<LiveProvider code={code} scope={{ ToggleField, useState }}>
  <LiveEditor style={livePreviewStyle}/>
  <LiveError />
  <LivePreview />
</LiveProvider>
<Wrapper />

The ToggleField is a Field component to be used in a Form, is composed of a [`Toggle`](../Toggle.mdx) and a `Label` and requires a `name` and `label`.

## Best practices

Any provided `id` will be automatically added to the label and toggle for accessibility; if an id is not provided, one will be created.

## Basic usage

```jsx
import React, { useState, useCallback } from 'react'
import { Button, ToggleField } from '@repay/cactus-web'

export default function Form() {
  const [values, setValues] = useState({ enabled: false })
  const handleChange = useCallback(
    (e) => {
      setValues(state => ({ ...state, [e.target.name]: e.target.checked }))
    },
    [setValues]
  )
  const onSubmit = useCallback(
    event => {
      event.preventDefault()
      const data = { ...values }
      // send data to api
    },
    [values]
  )

  return (
    <form onSubmit={onSubmit}>
      <ToggleField
        name="enabled"
        label="Is Enabled"
        onChange={handleChange}
        checked={values.enabled}
      />
      <Button type="submit" variant="action">
        Submit
      </Button>
    </form>
  )
}
```

View and edit the available props in the published [ToggleField stories](https://repaygithub.github.io/cactus/stories/cactus-web/?path=/story/togglefield--basic-usage) and review available props below.

## Properties

<PropsTable of={ToggleField} />
