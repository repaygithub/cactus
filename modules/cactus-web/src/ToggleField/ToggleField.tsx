import * as React from 'react'

import { FieldOnBlurHandler, FieldOnChangeHandler, FieldOnFocusHandler, Omit } from '../types'
import { margin, MarginProps } from 'styled-system'
import { omitMargins } from '../helpers/omit'
import FieldWrapper from '../FieldWrapper/FieldWrapper'
import handleEvent from '../helpers/eventHandler'
import Label, { LabelProps } from '../Label/Label'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Toggle, { ToggleProps } from '../Toggle/Toggle'
import useId from '../helpers/useId'

export interface ToggleFieldProps
  extends MarginProps,
    Omit<ToggleProps, 'id' | 'onChange' | 'onFocus' | 'onBlur'> {
  label: string
  /** props to be passed to the Label element (available for accessibility considerations) */
  labelProps?: Omit<LabelProps, 'children' | 'htmlFor'>
  id?: string
  /** required for form state management */
  name: string
  value?: boolean
  /** !important */
  disabled?: boolean
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void
  onChange?: FieldOnChangeHandler<boolean>
  onFocus?: FieldOnFocusHandler
  onBlur?: FieldOnBlurHandler
}

const ToggleFieldBase = (props: ToggleFieldProps) => {
  const {
    labelProps,
    className,
    label,
    id,
    name,
    onChange,
    onFocus,
    onBlur,
    onClick,
    ...toggleProps
  } = omitMargins(props) as Omit<ToggleFieldProps, keyof MarginProps>
  const fieldId = useId(id, name)

  const handleClick = React.useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      if (typeof onClick === 'function') {
        onClick(event)
      }
      if (typeof onChange === 'function') {
        const currentTarget = (event.currentTarget as unknown) as HTMLButtonElement
        onChange(name, currentTarget.getAttribute('aria-checked') !== 'true')
      }
    },
    [onClick, onChange, name]
  )

  const handleFocus = (event: React.FocusEvent) => {
    handleEvent(onFocus, name)
  }

  const handleBlur = (event: React.FocusEvent) => {
    handleEvent(onBlur, name)
  }

  return (
    <FieldWrapper className={className}>
      <Label {...props.labelProps} htmlFor={fieldId}>
        {props.label}
      </Label>
      <Toggle
        {...toggleProps}
        name={name}
        id={fieldId}
        onClick={handleClick}
        onFocus={handleFocus}
        onBlur={handleBlur}
      />
    </FieldWrapper>
  )
}

export const ToggleField = styled(ToggleFieldBase)`
  ${margin}

  ${Label} {
    cursor: pointer;
    margin-right: 8px;
  }

  ${Toggle} {
    vertical-align: text-bottom;
  }
`

// @ts-ignore
ToggleField.propTypes = {
  label: PropTypes.string.isRequired,
  labelProps: PropTypes.object,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  value: PropTypes.bool,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
}

ToggleField.defaultProps = {
  value: false,
}

export default ToggleField
