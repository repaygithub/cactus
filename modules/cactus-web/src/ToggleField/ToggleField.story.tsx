import { actions } from '@storybook/addon-actions'
import { boolean, text } from '@storybook/addon-knobs'
import { Meta } from '@storybook/react/types-6-0'
import React from 'react'

import FormHandler from '../storySupport/FormHandler'
import ToggleField from './ToggleField'

const eventLoggers = actions('onClick', 'onFocus', 'onBlur')

export default {
  title: 'ToggleField',
  component: ToggleField,
} as Meta

export const BasicUsage = (): React.ReactElement => (
  <FormHandler defaultValue={false} onChange={(name: string, value: boolean): boolean => value}>
    {({ value, onChange }): React.ReactElement => (
      <ToggleField
        name={text('name', 'boolean_field')}
        label={text('label', 'Boolean Field')}
        value={value}
        onChange={onChange}
        disabled={boolean('disabled', false)}
        {...eventLoggers}
      />
    )}
  </FormHandler>
)
