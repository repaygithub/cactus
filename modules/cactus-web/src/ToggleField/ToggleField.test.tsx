import { fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import ToggleField from './ToggleField'

describe('component: ToggleField', (): void => {
  test('snapshot', (): void => {
    const { container } = render(
      <StyleProvider>
        <ToggleField id="static-id" name="is_enabled" label="Enabled" value={false} />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('snapshot when value=true', (): void => {
    const { container } = render(
      <StyleProvider>
        <ToggleField id="static-id" name="is_enabled" label="Enabled" value={true} />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('snapshot when disabled', (): void => {
    const { container } = render(
      <StyleProvider>
        <ToggleField id="static-id" name="is_enabled" label="Enabled" value={false} disabled />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should generate unique id when one is not provided', (): void => {
    const { getByLabelText } = render(
      <StyleProvider>
        <ToggleField label="Show me the money" name="show-me-the-money" value={true} />
      </StyleProvider>
    )

    expect(getByLabelText('Show me the money').id).toContain('show-me-the-money')
  })

  test('should trigger onChange event with next value', (): void => {
    const onChange = jest.fn()
    const { getByLabelText } = render(
      <StyleProvider>
        <ToggleField
          label="Show me the money"
          name="show-me-the-money"
          value={true}
          onChange={onChange}
        />
      </StyleProvider>
    )

    fireEvent.click(getByLabelText('Show me the money'))
    expect(onChange).toHaveBeenCalledWith('show-me-the-money', false)
  })

  test('should trigger onFocus event', (): void => {
    const onFocus = jest.fn()
    const { getByLabelText } = render(
      <StyleProvider>
        <ToggleField label="Strange Nights" name="strange_nights" value={false} onFocus={onFocus} />
      </StyleProvider>
    )

    fireEvent.focus(getByLabelText('Strange Nights'))
    expect(onFocus).toHaveBeenCalled()
  })

  test('should trigger onBlur event', (): void => {
    const onBlur = jest.fn()
    const { getByLabelText } = render(
      <StyleProvider>
        <ToggleField label="Washed." name="washed" value={false} onBlur={onBlur} />
      </StyleProvider>
    )

    fireEvent.blur(getByLabelText('Washed.'))
    expect(onBlur).toHaveBeenCalled()
  })

  describe('when disabled', (): void => {
    test('should not trigger onChange event', (): void => {
      const onChange = jest.fn()
      const { getByLabelText } = render(
        <StyleProvider>
          <ToggleField label="Flow" name="flow" value={false} onChange={onChange} disabled />
        </StyleProvider>
      )

      userEvent.click(getByLabelText('Flow'))
      expect(onChange).not.toHaveBeenCalled()
    })

    test('should not trigger onFocus event', (): void => {
      const onFocus = jest.fn()
      const { getByLabelText } = render(
        <StyleProvider>
          <ToggleField
            label="Not For Sale"
            name="not_for_sale"
            value={true}
            onFocus={onFocus}
            disabled
          />
        </StyleProvider>
      )

      userEvent.click(getByLabelText('Not For Sale'))
      expect(onFocus).not.toHaveBeenCalled()
    })
  })
})
