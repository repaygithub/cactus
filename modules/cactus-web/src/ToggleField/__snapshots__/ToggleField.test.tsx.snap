// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ToggleField snapshot 1`] = `
.c2 + .c1 {
  margin-top: 16px;
}

.c4 {
  font-size: 18px;
  line-height: 28px;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c8 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 3px;
  left: 25px;
  color: hsl(0,0%,100%);
}

.c10 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 1px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c6 {
  position: relative;
  width: 45px;
  height: 20px;
  border-radius: 10px;
  outline: none;
  background-color: hsl(200,10%,20%);
  border: 1px solid hsl(200,10%,20%);
  cursor: pointer;
}

.c6:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c6::after {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  content: '';
  top: 0;
  left: 0;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
}

.c6[aria-checked='true'] {
  background-color: hsl(200,96%,35%);
  border-color: hsl(200,96%,35%);
}

.c6[aria-checked='true']::after {
  -webkit-transform: translateX(25px);
  -ms-transform: translateX(25px);
  transform: translateX(25px);
}

.c6[aria-checked='true'] .c7 {
  opacity: 0;
}

.c6[aria-checked='true'] .c9 {
  opacity: 1;
}

.c6 .c7 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6 .c9 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6[disabled] {
  background-color: hsl(0,0%,90%);
  border-color: hsl(0,0%,90%);
}

.c0 .c3 {
  cursor: pointer;
  margin-right: 8px;
}

.c0 .c5 {
  vertical-align: text-bottom;
}

<div>
  <div
    class="c0 c1 c2"
  >
    <label
      class="c3 c4"
      for="static-id"
    >
      Enabled
    </label>
    <button
      aria-checked="false"
      class="c5 c6"
      id="static-id"
      name="is_enabled"
      role="switch"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="c7 c8"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c9 c10"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`component: ToggleField snapshot when disabled 1`] = `
.c2 + .c1 {
  margin-top: 16px;
}

.c4 {
  font-size: 18px;
  line-height: 28px;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c8 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 3px;
  left: 25px;
  color: hsl(0,0%,100%);
}

.c10 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 1px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c11[aria-checked='true'] .c7 {
  opacity: 0;
}

.c11[aria-checked='true'] .c9 {
  opacity: 1;
}

.c11 .c7 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c11 .c9 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6 {
  position: relative;
  width: 45px;
  height: 20px;
  border-radius: 10px;
  outline: none;
  background-color: hsl(200,10%,20%);
  border: 1px solid hsl(200,10%,20%);
  cursor: cursor;
}

.c6:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c6::after {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  content: '';
  top: 0;
  left: 0;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
}

.c6[aria-checked='true'] {
  background-color: hsl(200,96%,35%);
  border-color: hsl(200,96%,35%);
}

.c6[aria-checked='true']::after {
  -webkit-transform: translateX(25px);
  -ms-transform: translateX(25px);
  transform: translateX(25px);
}

.c6[aria-checked='true'] .c7 {
  opacity: 0;
}

.c6[aria-checked='true'] .c9 {
  opacity: 1;
}

.c6 .c7 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6 .c9 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6[disabled] {
  background-color: hsl(0,0%,90%);
  border-color: hsl(0,0%,90%);
}

.c0 .c3 {
  cursor: pointer;
  margin-right: 8px;
}

.c0 .c5 {
  vertical-align: text-bottom;
}

<div>
  <div
    class="c0 c1 c2"
  >
    <label
      class="c3 c4"
      for="static-id"
    >
      Enabled
    </label>
    <button
      aria-checked="false"
      class="c5 c6"
      disabled=""
      id="static-id"
      name="is_enabled"
      role="switch"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="c7 c8"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c9 c10"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`component: ToggleField snapshot when value=true 1`] = `
.c2 + .c1 {
  margin-top: 16px;
}

.c4 {
  font-size: 18px;
  line-height: 28px;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c8 {
  vertical-align: middle;
  width: 12px;
  height: 12px;
  position: absolute;
  top: 3px;
  left: 25px;
  color: hsl(0,0%,100%);
}

.c10 {
  vertical-align: middle;
  width: 15px;
  height: 15px;
  position: absolute;
  top: 1px;
  left: 5px;
  color: hsl(0,0%,100%);
}

.c6 {
  position: relative;
  width: 45px;
  height: 20px;
  border-radius: 10px;
  outline: none;
  background-color: hsl(200,10%,20%);
  border: 1px solid hsl(200,10%,20%);
  cursor: pointer;
}

.c6:focus {
  box-shadow: 0 0 8px hsl(200,96%,35%);
}

.c6::after {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  content: '';
  top: 0;
  left: 0;
  position: absolute;
  -webkit-transition: -webkit-transform 0.3s;
  -webkit-transition: transform 0.3s;
  transition: transform 0.3s;
  background-color: hsl(0,0%,100%);
}

.c6[aria-checked='true'] {
  background-color: hsl(200,96%,35%);
  border-color: hsl(200,96%,35%);
}

.c6[aria-checked='true']::after {
  -webkit-transform: translateX(25px);
  -ms-transform: translateX(25px);
  transform: translateX(25px);
}

.c6[aria-checked='true'] .c7 {
  opacity: 0;
}

.c6[aria-checked='true'] .c9 {
  opacity: 1;
}

.c6 .c7 {
  opacity: 1;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6 .c9 {
  opacity: 0;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
}

.c6[disabled] {
  background-color: hsl(0,0%,90%);
  border-color: hsl(0,0%,90%);
}

.c0 .c3 {
  cursor: pointer;
  margin-right: 8px;
}

.c0 .c5 {
  vertical-align: text-bottom;
}

<div>
  <div
    class="c0 c1 c2"
  >
    <label
      class="c3 c4"
      for="static-id"
    >
      Enabled
    </label>
    <button
      aria-checked="true"
      class="c5 c6"
      id="static-id"
      name="is_enabled"
      role="switch"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="c7 c8"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M13.4142136,12 L22.7071068,21.2928932 C23.0976311,21.6834175 23.0976311,22.3165825 22.7071068,22.7071068 C22.3165825,23.0976311 21.6834175,23.0976311 21.2928932,22.7071068 L12,13.4142136 L2.70710678,22.7071068 C2.31658249,23.0976311 1.68341751,23.0976311 1.29289322,22.7071068 C0.902368927,22.3165825 0.902368927,21.6834175 1.29289322,21.2928932 L10.5857864,12 L1.29289322,2.70710678 C0.902368927,2.31658249 0.902368927,1.68341751 1.29289322,1.29289322 C1.68341751,0.902368927 2.31658249,0.902368927 2.70710678,1.29289322 L12,10.5857864 L21.2928932,1.29289322 C21.6834175,0.902368927 22.3165825,0.902368927 22.7071068,1.29289322 C23.0976311,1.68341751 23.0976311,2.31658249 22.7071068,2.70710678 L13.4142136,12 Z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c9 c10"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
        />
      </svg>
    </button>
  </div>
</div>
`;
